winml/test/concurrency/ConcurrencyTests.cpp

```diff
@@ -0,0 +1,279 @@
+{
+    WINML_SKIP_TEST("Skipping due to bug 21617097");
+
+    constexpr auto loadTestModel = [](const std::string& instance, LearningModelDeviceKind device)
+    {
+        WinML::Engine::Test::ModelValidator::SqueezeNet(instance, device, 0.00001f, false);
+    };
+
+    std::vector<std::thread> threads;
+    for (const auto& instance : {"1", "2", "3", "4"})
+    {
+        threads.emplace_back(loadTestModel, instance, LearningModelDeviceKind::Cpu);
+    }
+    if (GPUTEST_ENABLED) {
+        for (const auto& instance : {"GPU_1", "GPU_2", "GPU_3", "GPU_4"}) {
+            threads.emplace_back(loadTestModel, instance, LearningModelDeviceKind::DirectX);
+        }
+    }
+
+    for (auto& thread : threads) {
+        thread.join();
+    }
+}
+
+void ConcurrencyTestsApiSetup()
+{
+    init_apartment();
+    std::srand(static_cast<unsigned>(std::time(nullptr)));
+}
+
+struct EvaluationUnit
+{
+    LearningModel model;
+    LearningModelSession session;
+    LearningModelBinding binding;
+    winrt::Windows::Foundation::IAsyncOperation<LearningModelEvaluationResult> operation;
+    LearningModelEvaluationResult result;
+
+    EvaluationUnit() : model(nullptr), session(nullptr), binding(nullptr), result(nullptr) {}
+};
+
+// Run EvalAsync for each unit concurrently and get results
+static void RunAsync(std::vector<EvaluationUnit> &evalUnits)
+{
+    std::for_each(evalUnits.begin(), evalUnits.end(), [](EvaluationUnit &unit) {
+        unit.operation = unit.session.EvaluateAsync(unit.binding, L"");
+    });
+    // get results
+    std::for_each(evalUnits.begin(), evalUnits.end(), [](EvaluationUnit &unit) {
+        unit.result = unit.operation.get();
+    });
+}
+
+static void VerifyEvaluation(const std::vector<EvaluationUnit> &evalUnits, std::vector<uint32_t> expectedIndices)
+{
+    assert(evalUnits.size() == expectedIndices.size());
+    for (size_t i = 0; i < evalUnits.size(); ++i)
+    {
+        auto unit = evalUnits[i];
+        auto expectedIndex = expectedIndices[i];
+        auto result = unit.result.Outputs().Lookup(L"softmaxout_1").as<TensorFloat>().GetAsVectorView();
+        int64_t maxIndex = 0;
+        float maxValue = 0;
+        for (uint32_t j = 0; j < result.Size(); ++j)
+        {
+            float val = result.GetAt(j);
+            if (val > maxValue)
+            {
+                maxValue = val;
+                maxIndex = j;
+            }
+        }
+        WINML_EXPECT_TRUE(maxIndex == expectedIndex);
+    }
+}
+
+static void copyFile(LPCWSTR from, LPCWSTR to)
+{
+    using namespace std;
+    ifstream source(FileHelpers::GetModulePath() + from, ios::binary);
+    ofstream dest(FileHelpers::GetModulePath() + to, ios::binary);
+
+    dest << source.rdbuf();
+}
+
+// Run evaluations with different models
+void EvalAsyncDifferentModels()
+{
+    copyFile(L"model.onnx", L"model2.onnx");
+
+    std::vector<std::wstring> modelPaths = { L"model.onnx", L"model2.onnx" };
```

Comment:
I know this is a port, but at this point (or maybe shortly after if we cant do it now), we should conform these tests with the onnxruntime coding guidelines: snake_case.